### DISCUSSION TITLE
Block editor behaves unexpectedly...

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#106] [General]
Hi,

I am using the following setup: Raspi Pi Pico W, WiFi firmware loaded, additional code loaded, Wifi not tested yet. On Linux I use picocom. Zeptoforth (full big usb) version taken from the 1.5.6 release is flashed onto the Pico. So far it works nice.
I tried the block editor with block 10. Sometimes the entered text remains and survives even a powercycle, sometime it contains binary garbage after re-entering the same block with the block editor again.
What did I wrong?
What I want is a way to edit small forth programs directly on the Pico in a persistent way. How can I accomplish this?

**Comment by tabemann** [comment]
On Thu, Jun 27, 2024 at 11:28 AM mcctuxic ***@***.***> wrote:

> Hi,
>
> I am using the following setup: Raspi Pi Pico W, WiFi firmware loaded,
> additional code loaded, Wifi not tested yet. On Linux I use picocom.
> Zeptoforth (full big usb) version taken from the 1.5.6 release is flashed
> onto the Pico. So far it works nice.
> I tried the block editor with block 10. Sometimes the entered text remains
> and survives even a powercycle, sometime it contains binary garbage after
> re-entering the same block with the block editor again.
> What did I wrong?
> What I want is a way to edit small forth programs directly on the Pico in
> a persistent way. How can I accomplish this?
>
Thanks for catching this. I honestly have not used the block editor much on
the Pico or Pico W. I will look into it. (Note that the block editor does
not save blocks by default, unless you change blocks such as to result in
the block one had been editing being evicted. I assume that you know that
you have to save blocks explicitly under normal conditions.)

One thing, however, is that if you want to store code persistently, I would
recommend using an SDHC card with a FAT32 filesystem hooked up using SPI,
on which you will be able to edit and save files using zeptoed. (You have
to manually load zeptoed, it should be noted.) zeptoed is far more powerful
than the block editor that is provided with zeptoforth, and I have tested
it much more extensively. (Warning. though - it does not like it when you
make it run out of memory in the heap space it allocates.)

Travis

**Comment by tabemann** [comment]
On Thu, Jun 27, 2024 at 2:48 PM Travis Bemann ***@***.***> wrote:

> On Thu, Jun 27, 2024 at 11:28 AM mcctuxic ***@***.***>
> wrote:
>
>> Hi,
>>
>> I am using the following setup: Raspi Pi Pico W, WiFi firmware loaded,
>> additional code loaded, Wifi not tested yet. On Linux I use picocom.
>> Zeptoforth (full big usb) version taken from the 1.5.6 release is flashed
>> onto the Pico. So far it works nice.
>> I tried the block editor with block 10. Sometimes the entered text
>> remains and survives even a powercycle, sometime it contains binary garbage
>> after re-entering the same block with the block editor again.
>> What did I wrong?
>> What I want is a way to edit small forth programs directly on the Pico in
>> a persistent way. How can I accomplish this?
>>
> Thanks for catching this. I honestly have not used the block editor much
> on the Pico or Pico W. I will look into it. (Note that the block editor
> does not save blocks by default, unless you change blocks such as to result
> in the block one had been editing being evicted. I assume that you know
> that you have to save blocks explicitly under normal conditions.)
>
> One thing, however, is that if you want to store code persistently, I
> would recommend using an SDHC card with a FAT32 filesystem hooked up using
> SPI, on which you will be able to edit and save files using zeptoed. (You
> have to manually load zeptoed, it should be noted.) zeptoed is far more
> powerful than the block editor that is provided with zeptoforth, and I have
> tested it much more extensively. (Warning. though - it does not like it
> when you make it run out of memory in the heap space it allocates.)
>

I am having trouble reproducing the block corruption issue. I am using a
Pico W with a full big usb 1.5.6 image loaded along with the WiFi firmware,
and have been repeatedly editing block 10, along with other blocks, and
rebooting in between edits and saves, and have seen no data corruption. I
have also wiped the block storage with BLOCK::ERASE-ALL-BLOCKS just in case
previously existing data was preventing this issue from manifesting itself,
but I still have not been able to elicit the issue. Could you erase all
your blocks with BLOCK::ERASE-ALL-BLOCKS and then write down, step by step,
a sequence of actions that causes this issue to manifest itself, so I can
reproduce it?

Thanks,

Travis

**Comment by mcctuxic** [comment]
Hi Travis,
(no native speaker ahead... ;) )

thank you very much for your quick reply! :)

This weekend I have time to recreate what you have suggested.
I will report later, when I have something reliable...

Cheers! :)

**Comment by mcctuxic** [comment]
Hi Travis,

I did a BLOCK::ERASE-ALL-BLOCKS on the board in question,
entered some code into block 10, loaded the block, run
some words defined by the contents of the block, reedit
the block by erasing contents and adding contents, rebooted
the board.... and nothing, absolutely nothing. The block
remains rock/block :) solid.

I installed zeptoforth on a Seeed studio XIAO RP2040 and
BLOCK::ERASE-ALL-BLOCKS all block beforehand.

Same behaviour: No corruption...nothing. Block solid! :)

In hindsight the whole incident of corruption seems to
me like nothing more than vague glimpse of a glitch...
in the matrix? 

Thank you very much for the hint with BLOCK::ERASE-ALL-BLOCKS!

Cheers! Have a good weekend!

**Comment by tabemann** [comment]
On Sat, Jun 29, 2024 at 8:31 AM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> I did a BLOCK::ERASE-ALL-BLOCKS on the board in question,
> entered some code into block 10, loaded the block, run
> some words defined by the contents of the block, reedit
> the block by erasing contents and adding contents, rebooted
> the board.... and nothing, absolutely nothing. The block
> remains rock/block :) solid.
>
> I installed zeptoforth on a Seeed studio XIAO RP2040 and
> BLOCK::ERASE-ALL-BLOCKS all block beforehand.
>
> Same behaviour: No corruption...nothing. Block solid! :)
>
> In hindsight the whole incident of corruption seems to
> me like nothing more than vague glimpse of a glitch...
> in the matrix?
>
> Thank you very much for the hint with BLOCK::ERASE-ALL-BLOCKS!
>
> Cheers! Have a good weekend!


Cool. It does make me wonder, though, what was happening? Did the block
storage get corrupted somehow that was causing these problems? (And it
would be interesting to find out how.)

Travis

**Comment by mcctuxic** [comment]
Thank you for the update of zeptoforth! :)

I will keep an eye on this. I have updated to zeptoforth 
1.5.6.1 big full usb on my Pi Pico this morning. After that
I installed the firmware and ZeptoIP. Until now there is no
"glitch in the matrix"...
I will report back, if someting will happen.

Cheers!
Tuxic
